src/core/thread/device_mode.hpp

```diff
@@ -0,0 +1,236 @@
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for MLE device mode.
+ */
+
+#ifndef DEVICE_MODE_HPP_
+#define DEVICE_MODE_HPP_
+
+#include "openthread-core-config.h"
+
+#include <openthread/thread.h>
+
+#include "common/string.hpp"
+#include "utils/wrap_stdint.h"
+
+namespace ot {
+namespace Mle {
+
+/**
+ * @addtogroup core-mle-core
+ *
+ * @brief
+ *   This module includes definition for MLE device mode.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This type represents a MLE device mode.
+ *
+ */
+class DeviceMode
+{
+public:
+    enum
+    {
+        kModeRxOnWhenIdle      = 1 << 3, ///< If the device has its receiver on when not transmitting.
+        kModeSecureDataRequest = 1 << 2, ///< If the device uses link layer security for all data requests.
+        kModeFullThreadDevice  = 1 << 1, ///< If the device is an FTD.
+        kModeFullNetworkData   = 1 << 0, ///< If the device requires the full Network Data.
+
+        kInfoStringSize = 45, ///< String buffer size used for `ToString()`.
+    };
+
+    /**
+     * This type defines the fixed-length `String` object returned from `ToString()`.
+     *
+     */
+    typedef String<kInfoStringSize> InfoString;
+
+    /**
+     *  This structure represents an MLE Mode configuration.
+     *
+     */
+    typedef otLinkModeConfig ModeConfig;
+
+    /**
+     * This is the default constructor for `DeviceMode` object.
+     *
+     */
+    DeviceMode(void) {}
```

Comment:
Is this necessary to declare/define? Just let this be implicitly declared/defined?

Reply:
Yes. I think we need it. The compiler adds an implicitly-declared default constructor only if there are no other user-specified constructors. Since we have such constructors here, I think we need to explicitly add/define it.